peermanager/peerwantmanager.go

```diff
@@ -0,0 +1,226 @@
+			wantBlocks: cid.NewSet(),
+			wantHaves:  cid.NewSet(),
+		}
+	}
+}
+
+func (pwm *peerWantManager) RemovePeer(p peer.ID) {
+	delete(pwm.peerWants, p)
+}
+
+func (pwm *peerWantManager) BroadcastWantHaves(wantHaves []cid.Cid) map[peer.ID][]cid.Cid {
+	res := make(map[peer.ID][]cid.Cid)
+
+	// Iterate over all known peers
+	for p, pws := range pwm.peerWants {
+		// Iterate over all want-haves
+		for _, c := range wantHaves {
+			// If the CID has not been sent as a want-block or want-have
+			if !pws.wantBlocks.Has(c) && !pws.wantHaves.Has(c) {
+				// Record that the CID has been sent as a want-have
+				pws.wantHaves.Add(c)
+
+				// Add the CID to the results
+				if _, ok := res[p]; !ok {
+					res[p] = make([]cid.Cid, 0, 1)
+				}
+				res[p] = append(res[p], c)
+			}
+		}
+	}
+
+	return res
+}
+
+func (pwm *peerWantManager) SendWants(p peer.ID, wantBlocks []cid.Cid, wantHaves []cid.Cid) ([]cid.Cid, []cid.Cid) {
+	resWantBlks := make([]cid.Cid, 0)
+	resWantHvs := make([]cid.Cid, 0)
+
+	// Get the existing want-blocks and want-haves for the peer
+	if pws, ok := pwm.peerWants[p]; ok {
+		// Iterate over the requested want-blocks
+		for _, c := range wantBlocks {
+			// If the want-block hasn't been sent to the peer
+			if !pws.wantBlocks.Has(c) {
+				// Record that the CID was sent as a want-block
+				pws.wantBlocks.Add(c)
+
+				// Add the CID to the results
+				resWantBlks = append(resWantBlks, c)
+
+				// Make sure the CID is no longer recorded as a want-have
+				pws.wantHaves.Remove(c)
+
+				// Increment the count of want-blocks
+				pwm.wantBlockGauge.Inc()
+			}
+		}
+
+		// Iterate over the requested want-haves
+		for _, c := range wantHaves {
+			// If the CID has not been sent as a want-block or want-have
+			if !pws.wantBlocks.Has(c) && !pws.wantHaves.Has(c) {
+				// Record that the CID was sent as a want-have
+				pws.wantHaves.Add(c)
+
+				// Add the CID to the results
+				resWantHvs = append(resWantHvs, c)
+			}
+		}
+	}
+
+	return resWantBlks, resWantHvs
+}
+
+func (pwm *peerWantManager) SendCancels(cancelKs []cid.Cid) map[peer.ID][]cid.Cid {
+	res := make(map[peer.ID][]cid.Cid)
+
+	// Iterate over all known peers
+	for p, pws := range pwm.peerWants {
+		// Iterate over all requested cancels
+		for _, c := range cancelKs {
+			isWantBlock := pws.wantBlocks.Has(c)
+			isWantHave := pws.wantHaves.Has(c)
+
+			// If the CID was sent as a want-block, decrement the want-block count
+			if isWantBlock {
+				pwm.wantBlockGauge.Dec()
+			}
+
+			// If the CID was sent as a want-block or want-have
+			if isWantBlock || isWantHave {
+				// Remove the CID from the recorded want-blocks and want-haves
+				pws.wantBlocks.Remove(c)
+				pws.wantHaves.Remove(c)
+
+				// Add the CID to the results
+				if _, ok := res[p]; !ok {
+					res[p] = make([]cid.Cid, 0, 1)
+				}
+				res[p] = append(res[p], c)
+			}
+		}
+	}
+
+	return res
+}
+
+func (pwm *peerWantManager) PeerCanSendWants(p peer.ID, wants []cid.Cid) []cid.Cid {
```

Comment:
This looks like it's for testing. Let's describe what it does and make it clear that it's for testing.

Maybe don't export it?